homeassistant/scripts/auth.py

```diff
@@ -32,54 +35,114 @@ def run(args):
 
 
-async def list_users(data, args):
+async def list_users(auth_manager, data, args):
     """List the users."""
     count = 0
-    for user in data.users:
-        count += 1
-        print(user['username'])
+    if args.all:
+        # pylint: disable=protected-access
+        for user in await auth_manager._store.async_get_users():
+            print("{}{}{}".format(
+                str(user.name).ljust(20),
+                str(user.id).ljust(34),
+                str(user.mfa_modules)
+            ))
+            count += 1
+            for cred in user.credentials:
+                print("  - {}".format(cred.data.get('username')))
+
+    else:
+        provider = list(auth_manager.async_auth_providers)[0]
+        for user in await provider.async_credentials():
+            count += 1
+            print(user.data.get('username'))
 
     print()
     print("Total users:", count)
 
 
-async def add_user(data, args):
+async def add_user(auth_manager, data, args):
     """Create a user."""
     data.add_user(args.username, args.password)
     await data.async_save()
     print("User created")
 
 
-async def validate_login(data, args):
+async def validate_login(auth_manager, data, args):
     """Validate a login."""
     try:
         data.validate_login(args.username, args.password)
-        print("Auth valid")
-    except hass_auth.InvalidAuth:
+        if args.code:
+            provider = list(auth_manager.async_auth_providers)[0]
+            credential = await provider.async_get_or_create_credentials(
+                {'username': args.username})
+            user = await auth_manager.async_get_or_create_user(credential)
+
+            module = await auth_manager.async_get_auth_module('totp')
+            result = await module.async_validation_flow(
+                user.id, {'code': args.code})
+            if result is not None:
+                print("Auth valid")
+            else:
+                print("Auth invalid")
+        else:
+            print("Auth valid")
+    except (hass_auth.InvalidAuth, auth.InvalidAuth):
         print("Auth invalid")
 
 
-async def change_password(data, args):
+async def change_password(auth_manager, data, args):
     """Change password."""
     try:
         data.change_password(args.username, args.new_password)
         await data.async_save()
         print("Password changed")
     except hass_auth.InvalidUser:
         print("User not found")
+
+
+async def enable_mfa(auth_manager, data, args):
+    """Enable mfa for user."""
+    try:
+        data.validate_login(args.username, args.password)
+
+        provider = list(auth_manager.async_auth_providers)[0]
+        credential = await provider.async_get_or_create_credentials(
+            {'username': args.username})
+        user = await auth_manager.async_get_or_create_user(credential)
+        secret = await auth_manager.async_enable_user_mfa(user, 'totp')
+        # FIXME need to wait until AuthStore save finish
```

Comment:
@balloob I cannot find a way to wait AuthStore.async_save() actually done, since it will add a task delay 1 sec start.

Reply:
Try `await hass.async_stop()`

Reply:
Or `await asyncio.sleep(1.1)` 😉 (it's a script so fine with me)

Reply:
`await hass.async_stop()` works, leave two INFO log entries on console (stop event), but I think it is fine.